SF-813 - Lines are not wrapping correctly in texts

nigel-wells opened the pull request:
- Moved box-decoration-break from `usx-para-contents` to `usx-segment` and change to `clone`
- Add additional padding to `usx-segment` to display nicely when it wraps
- Changed `usx-verse` to be `display: inline-block` to ensure verses aren't forcing a line break when text is being added

I realise the change to `usx-verse` may seem like a regression as I believe the previous iteration was to ensure verses stayed on the same line as the contents of that verse - this change will stop that. In order to actually achieve that we will either need to encompass the verse and contents in in their own element or add the verse number as a `data` attribute on the `usx-segment` and then make use of `:before` or `:after` depending on which one is NOT currently used for the question icon. I suspect we may need to use `:before` or `:after` as modifying the HTML will affect what goes back to paratext won't it?

@Nateowami I've not been able to test this on Linux Firefox to see if these changes reintroduce the issues we solved before which is why this is a draft PR. Do you mind testing it out for me?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/web-xforge/644)
<!-- Reviewable:end -->

nigel-wells commented:
Working on this again now.
Just to clarify, there are two bugs here in this PR:
1. `box-decoration-break` which appears to be a bug in Firefox
2. Text being entered that is split across multiple lines has the text entered on a new line and, when a space is entered (and there is room on the previous line) the word gets moved to the previous line and the cursor remains on the second line. This may be related and fixed by the first point but the `inline-block` was the rational to resolve this issue. Typically if you want multiple elements to stay together then then need to be contained by another element otherwise they get rendered and placed independently. As the `usx-verse` and `usx-segment` are separated then I would expect them to wrap the way they are. I'll look in to this further.

nigel-wells commented:
After doing some more tests it looks like the moving of `box-decoration-break` from `usx-para-contents` to `usx-segment` has resolve the issue on Firefox. 
*`box-decoration-break: clone` on `usx-para-contents`*
![image](https://user-images.githubusercontent.com/17464863/82269295-590f3f80-99c5-11ea-86e4-61294882229b.png)

*`box-decoration-break: clone` on `usx-segment`*
![image](https://user-images.githubusercontent.com/17464863/82269306-675d5b80-99c5-11ea-810b-1bda08bb2af8.png)

I had hoped it might have just been an inheritance issue and it appears to be the case - at least from my limited testing anyway.

I see the `inline-block` adjustment creates an absolute mess after rebasing with `master` - I hope that wasn't what you guys were looking at when you tested!

nigel-wells commented:
The latest update fixes up the `usx-verse` issue that forced the adding of text on a second line when there was room for it to be on the first line.

I've tested in Chrome and Firefox on Windows 10 and the `box-decoration-break` issue appears resolved as well.

The last issue remaining is the `usx-verse` no wrapping with the `usx-segment` related to it. As mentioned above I do not believe this is going to be possible with the current DOM. In order for them to wrap together they need to be in the same container i.e.
```
<usx-segment-container>
  <usx-verse><span><span><span>1</span></span></span></usx-verse>
  <usx-segment>Text goes here</usx-segment>
</usx-segment-container>
```
I had considered having the verse number as a `data` attribute on the `usx-segment` and then making use of `:before` or `:after` but both of those have already been used by the checking app for displaying the question icon and count.
Are we safe to add in some additional elements for display purposes and then ensure they aren't synced back to paratext?

What is

nigel-wells commented:
I have found a way we could do it with using `:before` on the `usx-verse` but it would require JS to determine if the `usx-segment` has moved on to another line or not to know when to apply the CSS or not. We do some real time processing for working out the direction of the paragraph so we might be able to hook in to that.

ddaspit commented:
Regarding the first problem, moving `box-decoration-break` to `usx-segment` adds a lot of extra padding around every segment. For example, the small blank segment at the beginning of a paragraph is now rather large.

Regarding the second problem, I was able to fix it without using `inline-block` at all.  In the `create` method of the `VerseEmbed` class, I added a `wbr` element instead of a zero-width space and it fixed the problem. Replace
```
node.appendChild(document.createTextNode(ZWSP));
```
with
```
node.appendChild(document.createElement('wbr'));
```

nigel-wells commented:
Spacing can easily be adjusted with CSS - I'm not convinced we should be using HTML to create space in elements as that isn't its purpose.
I've tried added in the `wbr` element but it appears to have no affect - there is no CSS associated with it and it sits inside the `usx-verse` element. This doesn't help with ensuring the `usx-verse` element wraps with the `usx-segment` element. Did you do something else as well?

nigel-wells commented:
I've pushed through where I got up to. The CSS is fine but for some reason quill isn't adding the class name to `usx-verse` and instead creates a container around it. Is this to do with how the `usx-verse` is setup as I don't remember having this issue setting up the question `data` attributes.

ddaspit commented:
Did you remove `inline-block` from `usx-verse`?

ddaspit commented:
The problem with moving `box-decoration-break` to `usx-segment` and adding padding is that you will now have extra padding at the beginning and end of every segment that doesn't need to be there just so that you can get padding at the line breaks. There is no way to get rid of the undesirable padding with CSS. When `box-decoration-break` was set to `clone` on `usx-para-contents`, then there was only padding at the beginning and end of the paragraph (and at the line breaks), which is what we want. Is there anyway to duplicate that behavior with `box-decoration-break` set on `usx-segment`?

ddaspit commented:
Check out the `bug/SF-813a` branch that I pushed. It fixes the first problem using a `wbr` element.

nigel-wells commented:
Well hot dog - never even heard of `<wbr>` as being a valid HTML element! It's even compatible way back to IE 5.5! This is a much better way - learn something every day :) I'll take a look at tweaking the spacing tomorrow - I'm sure we can come up with something good.

ddaspit commented:
Since the second issue is caused by a bug in Firefox (https://bugzilla.mozilla.org/show_bug.cgi?id=1569107), I think we should use `box-decoration-break: clone` on all other browsers. I pushed the `bug/SF-813b` branch to demonstrate. If we can come up with a workaround, then we will only need to apply it on Firefox.

ddaspit commented:
On Firefox, I also noticed that the `box-shadow` that is applied to `usx-para-contents` when it is selected seems to be weird when there is RTL text.
![rtl-firefox-drop-shadow](https://user-images.githubusercontent.com/3261883/82346186-5b52b780-9a20-11ea-98f8-04822c128e41.png)

nigel-wells commented:
The `box-shadow` issue seems related to the same Firefox bug. I was able to "resolve" it by increasing the shadow a little more to fill the space and then moving it up so that it looks like it takes the same amount of space as before. I also added the `<wbr>` in to the `bug/SF-813b` branch. 

I think everything is now addressed in this branch?

The only other thought I had was whether it would be beneficial to refactor your `isFirefox()` to be more generic so that quill simply stored a class for whatever browser was being used? Otherwise we should have a note somewhere explaining why Quill needs to know about Firefox and no one else.

nigel-wells commented:
Closing in favour of #661
